lib/api/src/entities/function/host/imp.rs

```diff
@@ -0,0 +1,730 @@
+             let args_without_store: JSArray = arguments[1..].try_into().unwrap();
+             let [ $( $x ),* ] = args_without_store;
+             let mut store = StoreMut::from_raw(store_ptr as usize as *mut _);
+             func(RuntimeFunctionEnvMut::Jsc(env).into(), $( FromToNativeWasmType::from_native( $x::Native::from_raw(&mut store, RawValue { u128: {
+                 // TODO: This may not be the fastest way, but JSC doesn't expose a BigInt interface
+                 // so the only thing we can do is parse from the string repr
+                 if $x.is_number(&ctx) {
+                     $x.to_number(&ctx).unwrap() as _
+                 }
+                 else {
+                     $x.to_string(&ctx).unwrap().to_string().parse::<u128>().unwrap()
+                 }
+             } }) ) ),* ).into_result()
+         }));
+
+         match result {
+             Ok(Ok(result)) => {
+                 match Rets::size() {
+                     0 => {Ok(JSValue::undefined(&ctx))},
+                     1 => {
+                         // unimplemented!();
+
+                         let ty = Rets::wasm_types()[0];
+                         let mut arr = result.into_array(&mut store);
+                         // Value::from_raw(&store, ty, arr[0])
+                         let val = Value::from_raw(&mut store, ty, arr.as_mut()[0]);
+                         let value: JSValue = val.as_jsc_value(&store);
+                         Ok(value)
+                         // *mut_rets = val.as_raw(&mut store);
+                     }
+                     _n => {
+                         // if !results.is_array(&context) {
+                         //     panic!("Expected results to be an array.")
+                         // }
+                         let mut arr = result.into_array(&mut store);
+                         let result_values = Rets::wasm_types().iter().enumerate().map(|(i, ret_type)| {
+                             let raw = arr.as_mut()[i];
+                             Value::from_raw(&mut store, *ret_type, raw).as_jsc_value(&mut store)
+                         }).collect::<Vec<_>>();
+                         Ok(JSObject::new_array(&ctx, &result_values).unwrap().to_jsvalue())
+                     }
+                 }
+             },
+             #[cfg(feature = "std")]
+             Ok(Err(err)) => {
+                 let trap = crate::jsc::vm::Trap::user(Box::new(err));
+                 Err(trap.into_jsc_value(&ctx))
+             },
+             #[cfg(feature = "core")]
+             Ok(Err(err)) => {
+                 let trap = crate::jsc::vm::Trap::user(Box::new(err));
+                 Err(trap.into_jsc_value(&ctx))
+             },
+             Err(panic) => {
+                 Err(JSValue::string(&ctx, format!("panic: {:?}", panic)))
+             },
+         }
+
+     }
+
+     Some(fn_callback::<T, $( $x, )* Rets, RetsAsResult, Self > as _)
+   }
+
+  #[cfg(feature = "js")]
```

Comment:
These impulse are quite a lot of code. 

Could they life in the backend-specific modules instead?

Reply:
[81f4195](https://github.com/wasmerio/wasmer/pull/5243/commits/81f41957f94a32924733020ad2373d0158fe60d1)